### PR TITLE
[DO NOT MERGE] rely on native Kafka headers (wip)

### PR DESCRIPTION
**Pull request by markfisher**
I was trying to get this working along with the changes in the "sync-request-reply-boot2" branch of sk8s. Not quite there yet (I'm seeing no activity in the sidecar log at all).

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -20,3 +20,6 @@
 #  name = "github.com/x/y"
 #  version = "2.4.0"
 
+[[override]]
```
Comment (ericbottard) [inline Gopkg.toml]: According to https://github.com/golang/dep/blob/master/docs/Gopkg.toml.md, `constraint` should be sufficient, shouldn't it?

File: cmd/function-sidecar.go [hunk 2]
```diff
@@ -80,27 +79,19 @@ func main() {
...
 				if output != nil {
-					messageOut := message.Message{Payload: []byte(dispatched.(string)), Headers: messageIn.Headers}
-					bytesOut, err := message.EncodeMessage(messageOut)
-					fmt.Fprintf(os.Stdout, ">>> %s\n", messageOut)
-					if err != nil {
-						log.Printf("Error encoding message: %v", err)
-						break
+					bytesOut := []byte(dispatched.(string));
+					replyHeaders := make([]sarama.RecordHeader, len(msg.Headers))
+					for i, h := range msg.Headers {
```
Comment (ericbottard) [inline cmd/function-sidecar.go]: Should be able to use the builtin `copy`, if not pass the incoming slice directly (I assume you wanted to make a semantic distinction between the two?)

**Comment by ericbottard** [comment]
> I was trying to get this working along with the changes in the "sync-request-reply-boot2" branch of sk8s. Not quite there yet (I'm seeing no activity in the sidecar log at all).

My understanding was that this work should work on its own

File: Gopkg.toml (hunk 1 above)
Comment (markfisher) [reply Gopkg.toml]: Okay. This is new territory for me.

File: cmd/function-sidecar.go (hunk 2 above)
Comment (markfisher) [reply cmd/function-sidecar.go]: I agree there should be a simpler way, but after trying several things, this was all I got to work for now. That said, it's clearly just my misunderstanding of how dereferencing works with arrays/slices at this point (including when it's handled automatically). Here's what I tried first (passing the inbound headers as-is):

```
outMessage := &sarama.ProducerMessage{Topic: output.(string), Value: sarama.ByteEncoder(bytesOut), Headers: msg.Headers}
```

And that led to this error from `make dockerize`:

```
cmd/function-sidecar.go:94:112: cannot use msg.Headers (type []*sarama.RecordHeader) as type []sarama.RecordHeader in field value
```

I also tried dereferencing at the slice level, but that led to syntax errors.

Then I tried `copy(replyHeaders, msg.Headers)`, but received a similar build error:

```
cmd/function-sidecar.go:91:10: arguments to copy have different element types: []sarama.RecordHeader and []*sarama.RecordHeader
```

Again, I experimented with dereferencing syntax and a couple different ways of declaring `replyHeaders` but I ended up with the range and item-by-item pointer copy as something that worked. I'd be happy to learn how to handle this as closely as possible to the direct pass-thru I tried above (working on that learning part now, but feel free to share some knowledge ;-)).

Comment (ericbottard) [reply cmd/function-sidecar.go]: Ah nevermind, I did not even see that there was de-referencing involved. golang auto accepting pointers for values is still a bit too magical to me.

File: Gopkg.toml (hunk 1 above)
Comment (ericbottard) [reply Gopkg.toml]: Indeed, tested with `constraint` and `dep ensure` puts the same stuff in `vendor/`

**Pull request closed without merging**